server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.storage.user;
+
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.UserModel;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This is an optional capability interface that is intended to be implemented by any
+ * <code>UserStorageProvider</code> that supports complex user querying. You must
+ * implement this interface if you want to view and manage users from the administration console.
+ * <p/>
+ * Note that all methods in this interface should limit search only to data available within the storage that is
+ * represented by this provider. They should not lookup other storage providers for additional information.
+ *
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public interface UserQueryMethods {
+    /**
+     * Searches all users in the realm.
+     *
+     * @param realm a reference to the realm.
+     * @return a non-null {@link Stream} of users.
+     * @deprecated Use {@link #searchForUserStream(RealmModel, Map)} with an empty params map instead.
+     */
+    @Deprecated
+    default Stream<UserModel> getUsersStream(RealmModel realm) {
+        return searchForUserStream(realm, Collections.emptyMap());
+    }
```

Comment:
Should the deprecated methods be kept in the new interface or rather remove them?

Reply:
I'd like to keep it here for the sake of this PR and remove deprecated methods in separate issue (https://github.com/keycloak/keycloak/issues/20438), which also has milestone 22.0.0. There is draft https://github.com/keycloak/keycloak/pull/20440 sent.